modules/core/src/main/java/org/apache/ignite/spi/ExponentialBackoffTimeoutStrategy.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.spi;
+
+import org.apache.ignite.internal.util.typedef.internal.S;
+import org.apache.ignite.internal.util.typedef.internal.U;
+
+/**
+ * Strategy which incorporates retriable network operation, handling of totalTimeout logic.
+ * It increases startTimeout based on exponential backoff algorithm.
+ *
+ * If failure detection is enabled it relies on totalTimeout
+ * otherwise implements exponential backoff totalTimeout logic based on startTimeout, maxTimeout and retryCnt.
+ */
+public class ExponentialBackoffTimeoutStrategy implements TimeoutStrategy {
+    /** Default backoff coefficient to calculate next timeout based on backoff strategy. */
+    private static final double DLFT_BACKOFF_COEFF = 2.0;
+
+    /** Max timeout of the next try, ms. */
+    private final long maxTimeout;
+
+    /** Total timeout, ms. */
+    private final long totalTimeout;
+
+    /** Timestamp of operation start to check totalTimeout. */
+    private final long start;
+
+    /** Current calculated timeout, ms. */
+    private long currTimeout;
+
+    /**
+     * Compute expected max backoff timeout based on initTimeout, maxTimeout and reconCnt and backoff coefficient.
+     *
+     * @param initTimeout Initial timeout.
+     * @param maxTimeout Max Timeout per retry.
+     * @param reconCnt Reconnection count.
+     * @return Calculated total backoff timeout.
+     */
+    public static long totalBackoffTimeout(
```

Comment:
I do not like static method here. Please implement additional method on interface level named maxWaitTime which will calculate max wait time using strategy specific for implementation.